keps/sig-testing/20190605-presubmitconfig-in-tested-repo.md

```diff
@@ -0,0 +1,259 @@
+  package
+* Posts a comment to GitHub if an error occurs
+* Updates the status context to `failure` or `success` after the
+  `prow.yaml` parsing is done
+* Has functionality to remove status contexts that existed in an older
+  `baseRef` and got removed in a newer one. This could be done by:
+    * Recognizing status contexts created by Prow  using a combination
+      of a static list (e.G. `tide`) and for the remaining contexts by
+      comparing their `targetURL` with what is configured in Prow as
+      `jobURLPrefix`. If the `targetURL` starts with `jobURlPrefix`
+      and the context does not belong to one of the known presubmits,
+      the status context will be retired.
+    * Leveraging the `status-reconciler`. The `status-reconciler`
+      is edge-driven, meaning whenever there is a configuration change
+      it will use the delta to find presubmits that got removed, then
+      check on all pull requests if they contain a status context for
+      a removed presubmit. To make this work for `prow.yaml`, we would
+      have to extend the `status-reconciler` to get an event from GitHub
+      whenever there is a pushevent to the base branch of any open
+      pull request, so that it can find out if a presubmit defined in
+      `prow.yaml` got removed
+
+
+*Alternative:* This could be put into the `prow/plugins/trigger` package
+ as that is the only consumer for this functionality
+
+#### Extend `prow/config` with configuration to opt-in to this feature
+
+This feature should be opt-in to avoid surprises. To achieve that, the
+`ProwConfig` struct in the `prow/config` package has to be extend with
+a `InRepoConfig     map[string]InRepoConfig` property. Also, a
+`func (pc *ProwConfig) InRepoConfigFor(org, repo string) InRepoConfig`
+has to be added that can be used to determine if this functionality is
+enabled. Initially, the `InRepoConfig` struct will be defined as follows:
+
+```
+type InRepoConfig struct {
+  Enabled bool `json:"enabled,omitempty"`
+}
+```
+
+This allows to add further configuration options to this feature in the
+future if needed.
+
+#### Add code to `tide` and the `trigger` plugin so they will use presubmits from a `prow.yaml` if the feature is activated
+
+Both `tide` and the `trigger` plugin will need some code that checks
+if the `prow.yaml` feature is activated and if yes, amend the presubmits
+from it to the presubmits that are configured in the `infra-config` repo.
+
+In `tide` this is very simple, we can just use the `prow/inrepoconfig/api` package to figure out the additional presubmits.
+
+*Open question:* What happens if the merged `prow.yaml` becomes invalid?
+Is tide able to recover from that?
+
+In the `trigger` plugin, we have to check if the `prow.yaml` feature is
+enabled and if yes, use the functionality described from the new `prow/inrepoconfig`
+package to fetch additional presubmits from it.
+
+
+### Risks and Mitigations
```

Comment:
@cjwagner Re your [comment:](https://github.com/kubernetes/test-infra/pull/12836#issuecomment-498373618)

> I'm excited to see this functionality, but I'm worried that this is going to cause problems by breaking assumptions.
> * It only works for job config not core prow config or plugin config.

Yes, that is true. What assumption exactly is broken by that?

> * Jobs defined in prow.yaml will be handled differently than jobs in the --job-config directory (prow.yaml jobs are parsed with the inclusion of PR changes).

By that you mean plugins/components that have a `ref` will or may get a different view than those that don't, like e.G. `branchprotector`?